wallet/tests/data_requests.rs

```diff
@@ -0,0 +1,159 @@
+#[actix_rt::test]
+async fn test_data_request_report_json_serialization() {
+    use witnet_data_structures::chain::{RADAggregate, RADRequest, RADRetrieve, RADTally, RADType};
+    use witnet_rad::{
+        script::{unpack_radon_script, RadonScriptExecutionSettings},
+        try_data_request,
+    };
+
+    let request = RADRequest {
+        time_lock: 0,
+        retrieve: vec![
+            RADRetrieve {
+                kind: RADType::HttpGet,
+                url: String::from("https://www.bitstamp.net/api/ticker/"),
+                script: vec![130, 24, 119, 130, 24, 100, 100, 108, 97, 115, 116],
+            },
+            RADRetrieve {
+                kind: RADType::HttpGet,
+                url: String::from("https://api.coindesk.com/v1/bpi/currentprice.json"),
+                script: vec![
+                    132, 24, 119, 130, 24, 102, 99, 98, 112, 105, 130, 24, 102, 99, 85, 83, 68,
+                    130, 24, 100, 106, 114, 97, 116, 101, 95, 102, 108, 111, 97, 116,
+                ],
+            },
+        ],
+        aggregate: RADAggregate {
+            filters: vec![],
+            reducer: 3,
+        },
+        tally: RADTally {
+            filters: vec![],
+            reducer: 3,
+        },
+    };
+
+    let inputs = [
+        r#"{"high": "9897.46000000", "last": "9723.56", "timestamp": "1591720963", "bid": "9717.67", "vwap": "9711.68", "volume": "6279.09256801", "low": "9566.81000000", "ask": "9723.56", "open": 9786.64}"#,
+        r#"{"time":{"updated":"Jun 9, 2020 16:42:00 UTC","updatedISO":"2020-06-09T16:42:00+00:00","updateduk":"Jun 9, 2020 at 17:42 BST"},"disclaimer":"This data was produced from the CoinDesk Bitcoin Price Index (USD). Non-USD currency data converted using hourly conversion rate from openexchangerates.org","chartName":"Bitcoin","bpi":{"USD":{"code":"USD","symbol":"&#36;","rate":"9,724.8354","description":"United States Dollar","rate_float":9724.8354},"GBP":{"code":"GBP","symbol":"&pound;","rate":"7,692.6949","description":"British Pound Sterling","rate_float":7692.6949},"EUR":{"code":"EUR","symbol":"&euro;","rate":"8,635.1092","description":"Euro","rate_float":8635.1092}}}"#,
+    ];
+    let report = try_data_request(
+        &request,
+        RadonScriptExecutionSettings::enable_all(),
+        Some(&inputs),
+    )
+    .await;
+    let aggregate_report = report.aggregate.clone();
+    let tally_report = report.tally.clone();
+
+    // Number of retrieval reports should match number of sources
+    assert_eq!(&report.retrieve.len(), &request.retrieve.len());
+
+    for (index, retrieve_report) in report.retrieve.iter().enumerate() {
+        // Each retrieval result must match last item in each retrieval partial results
+        assert_eq!(
+            &retrieve_report.result,
+            retrieve_report
+                .partial_results
+                .clone()
+                .unwrap()
+                .last()
+                .unwrap()
```

Comment:
Clone bad, as_ref good

```suggestion
                .as_ref()
                .unwrap()
                .last()
                .as_ref() // maybe this as_ref is not needed
                .unwrap()
```